docs/language-reference/07-declarations.md

```diff
@@ -0,0 +1,771 @@
+Declarations
+============
+
+Modules
+-------
+
+A module consists of one or more source units that are compiled together.
+The global declarations in those source units comprise the body of the module.
+
+In general, the order of declarations within a source unit does not matter; declarations can refer to other decalrations (of types, functions, variables, etc.) later in the same source unit.
```

Comment:
Except for our current parser disambiguation via knowing a kind.

Reply:
Yes, our current parser does some disambiguation based on being able to tell types and values apart.

The correct fix there is not to change the language reference (although a warning here would be nice), but rather to fix our parser so that the bodies of functions/structures/etc. don't get parsed at first (instead we just capture all the tokens between matching `{}` and save them for later). Then we can go back later to parse the bodies of things *after* we've built up the full global scope so that forward references can be resolved.

The trend in stuff like Rust and Swift is in the other direction of making sure that a context-free parse of the entire input is possible without any kind of type information, while still supporting out of order declarations. The argument is that allowing for a context-free parse makes it easier to write stand-alone tools for documentation generation, refactoring, highlighting, etc.

Those languages aren't wrong about the benefits, but we don't have a choice given that we've inherited C-like syntax with its ambiguities, and dxc has added some C++ ambiguities to their flavor of HLSL creating even more potential problems down the line.

One down-side to the "it must be possible to parse without knowing anything" rule is that it means that invocations of user-defined syntax macros need to be marked (e.g., `!someMacro` in Rust) *and* the bodies of macro invocations need to be handled as completely undifferentiated token streams by tools (significantly impairing their utility in macro-heavy code, which some Rust projects, including most that try to do anything with GPUs/graphics).